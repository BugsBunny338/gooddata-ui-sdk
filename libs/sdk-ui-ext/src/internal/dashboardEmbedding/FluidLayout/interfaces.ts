// (C) 2019-2020 GoodData Corporation
import { Col, ColProps, Row, RowProps } from "react-grid-system";
import { IFluidLayoutColumn, IFluidLayoutRow, ResponsiveScreenType } from "@gooddata/sdk-backend-spi";

/**
 * Default props provided to a fluid layout row components and component callbacks.
 *
 * @alpha
 */
export type IFluidLayoutRowRenderProps<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>
> = {
    /**
     * Fluid layout row.
     */
    row: TRow;

    /**
     * Fluid layout row index.
     */
    rowIndex: number;

    /**
     * Current screen type with respect to the set breakpoints.
     */
    screen: ResponsiveScreenType;
};

/**
 * Fluid layout row key getter.
 * This callback is used to determine a unique key of the row.
 * By this callback, you can avoid unnecessary re-renders of the row components,
 * the returned unique key is passed to the React "key" property, when rendering rows.
 * By default, fluid layout will use rowIndex as a unique key.
 *
 * @alpha
 */
export type IFluidLayoutRowKeyGetter<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>
> = (props: IFluidLayoutRowRenderProps<TContent, TColumn, TRow>) => string;

/**
 * Fluid layout row renderer.
 * Represents a component for rendering the row.
 *
 * @alpha
 */
export type IFluidLayoutRowRenderer<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>,
    TCustomProps = object
> = React.ComponentType<
    IFluidLayoutRowRenderProps<TContent, TColumn, TRow> &
        RowProps & { ref?: React.RefObject<Row> } & TCustomProps
>;

/**
 * Default props provided to a fluid layout column components and component callbacks.
 *
 * @alpha
 */
export type IFluidLayoutColumnRenderProps<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>
> = {
    /**
     * Fluid layout row.
     */
    row: TRow;

    /**
     * Fluid layout row index.
     */
    rowIndex: number;

    /**
     * Fluid layout column.
     */
    column: TColumn;

    /**
     * Fluid layout column index.
     */
    columnIndex: number;

    /**
     * Current screen type with respect to the set breakpoints.
     */
    screen: ResponsiveScreenType;
};

/**
 * Fluid layout column key getter.
 * This callback is used to determine a unique key of the column.
 * By this callback, you can avoid unnecessary re-renders of the column components,
 * the returned unique key is passed to the React "key" property, when rendering columns.
 * By default, fluid layout will use columnIndex as a unique key.
 *
 * @alpha
 */
export type IFluidLayoutColumnKeyGetter<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>
> = (props: IFluidLayoutColumnRenderProps<TContent, TColumn, TRow>) => string;

/**
 * Fluid layout column renderer.
 * Represents a component for rendering the column.
 *
 * @alpha
 */
export type IFluidLayoutColumnRenderer<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>,
    TCustomProps = object
> = React.ComponentType<
    IFluidLayoutColumnRenderProps<TContent, TColumn, TRow> &
        ColProps & { ref?: React.RefObject<Col> } & TCustomProps
>;

/**
 * Fluid layout content renderer.
 * Represents a component for rendering the column content.
 *
 * @alpha
 */
export type IFluidLayoutContentRenderer<
    TContent,
    TColumn extends IFluidLayoutColumn<TContent>,
    TRow extends IFluidLayoutRow<TContent, TColumn>,
    TCustomProps = object
> = React.ComponentType<IFluidLayoutColumnRenderProps<TContent, TColumn, TRow> & TCustomProps>;
