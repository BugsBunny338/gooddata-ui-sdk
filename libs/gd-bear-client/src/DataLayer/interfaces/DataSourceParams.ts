// (C) 2007-2018 GoodData Corporation
import { AFM } from "@gooddata/gd-bear-model";

export interface IDataSourceParams {
    uri: string;
    attributeFilters?: AFM.AttributeFilterItem[];
    dateFilter?: AFM.DateFilterItem;
}
