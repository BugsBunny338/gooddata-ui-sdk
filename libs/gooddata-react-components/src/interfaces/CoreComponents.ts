// (C) 2019 GoodData Corporation
import * as React from "react";

import { IBaseChartProps, IChartProps } from "../_defunct/to_delete/BaseChart";
import { ITableProps } from "../components/core/PureTable";
import { ICommonVisualizationProps } from "../_defunct/to_delete/VisualizationLoadingHOC";
import { IDataSourceProviderInjectedProps } from "../_defunct/to_delete/DataSourceProvider";
import { IPivotTableProps } from "../_defunct/pivotTable/CorePivotTable";

export interface ICoreComponents {
    BaseChart: React.ComponentClass<IBaseChartProps>;
    Headline: React.ComponentClass<ICommonVisualizationProps & IDataSourceProviderInjectedProps>;
    Table: React.ComponentClass<ITableProps & IDataSourceProviderInjectedProps>;
    PivotTable: React.ComponentClass<IPivotTableProps>;
    ScatterPlot: React.ComponentClass<IChartProps>;
    FunnelChart: React.ComponentClass<IChartProps>;
}
