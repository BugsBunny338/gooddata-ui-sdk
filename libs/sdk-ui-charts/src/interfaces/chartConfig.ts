// (C) 2020 GoodData Corporation
import { ISeparators } from "@gooddata/numberjs";
import { IColor, IColorPalette } from "@gooddata/sdk-model";
import { IHeaderPredicate, VisType } from "@gooddata/sdk-ui";

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IChartConfig {
    type?: VisType;
    colors?: string[];
    colorPalette?: IColorPalette;
    colorMapping?: IColorMapping[];
    legend?: ILegendConfig;
    legendLayout?: string;
    limits?: IChartLimits;
    grid?: any;
    yFormat?: string;
    yLabel?: string;
    xFormat?: string;
    xLabel?: string;
    chart?: any;
    xaxis?: IAxisConfig;
    yaxis?: IAxisConfig;
    secondary_xaxis?: IAxisConfig;
    secondary_yaxis?: IAxisConfig;
    separators?: ISeparators;
    dataLabels?: IDataLabelsConfig;
    dualAxis?: boolean;
    primaryChartType?: string;
    secondaryChartType?: string;
    forceDisableDrillOnAxes?: boolean;
    disableDrillUnderline?: boolean;
    stacking?: boolean;
    stackMeasures?: boolean;
    stackMeasuresToPercent?: boolean;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export type PositionType = "left" | "right" | "top" | "bottom" | "auto";

/**
 * TODO: SDK8: add docs
 * @public
 */
export type IDataLabelsVisible = string | boolean;

/**
 * TODO: SDK8: add docs
 * @public
 */
export type ChartAlignTypes = "top" | "bottom" | "middle";

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IDataLabelsConfig {
    visible?: IDataLabelsVisible;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface ILegendConfig {
    enabled?: boolean;
    position?: PositionType;
    responsive?: boolean;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IChartLimits {
    series?: number;
    categories?: number;
    dataPoints?: number;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IColorMapping {
    // sent to SDK
    predicate: IHeaderPredicate;
    color: IColor;
}

export type AxisNamePosition = "high" | "low" | "middle";

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IAxisNameConfig {
    visible?: boolean;
    position?: AxisNamePosition;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IAxisConfig {
    visible?: boolean;
    labelsEnabled?: boolean;
    rotation?: string;
    min?: string;
    max?: string;
    measures?: string[];
    name?: IAxisNameConfig;
}
