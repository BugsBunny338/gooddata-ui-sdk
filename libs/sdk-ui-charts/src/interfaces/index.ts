// (C) 2007-2020 GoodData Corporation
export {
    IBucketChartProps,
    IChartCallbacks,
    ICommonChartProps,
    ICoreChartProps,
    OnLegendReady,
    ILegendData,
    ILegendItem,
} from "./chartProps";

export { ViewByAttributesLimit } from "./limits";
