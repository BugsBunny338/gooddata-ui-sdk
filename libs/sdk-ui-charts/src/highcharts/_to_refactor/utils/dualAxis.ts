// (C) 2019-2020 GoodData Corporation
import { measureLocalId } from "@gooddata/sdk-model";
import { IChartConfig } from "../../../interfaces";
import { BucketNames, VisType, DataViewFacade } from "@gooddata/sdk-ui";
import { isComboChart } from "./common";
import get from "lodash/get";

export function setMeasuresToSecondaryAxis(config: IChartConfig = {}, dv: DataViewFacade): IChartConfig {
    const isDualAxis: boolean = get(config, "dualAxis", true);
    const type: VisType = get(config, "type");
    const { secondary_yaxis: secondaryYAxis, ...remainConfig } = config;
    const secondaryMeasuresIds = dv.def().bucketMeasures(BucketNames.SECONDARY_MEASURES).map(measureLocalId);

    if (!isComboChart(type)) {
        return config;
    }

    if (!isDualAxis) {
        return remainConfig;
    }

    return {
        ...remainConfig,
        secondary_yaxis: {
            ...secondaryYAxis,
            measures: secondaryMeasuresIds,
        },
    };
}
