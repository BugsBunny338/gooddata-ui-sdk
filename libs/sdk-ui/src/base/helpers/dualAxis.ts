// (C) 2019 GoodData Corporation
import { DataViewFacade } from "@gooddata/sdk-backend-spi";
import { measureId } from "@gooddata/sdk-model";
import { IChartConfig } from "../interfaces/Config";
import { SECONDARY_MEASURES } from "../constants/bucketNames";
import { VisType } from "../constants/visualizationTypes";
import { isComboChart } from "./common";
import get = require("lodash/get");

export function setMeasuresToSecondaryAxis(config: IChartConfig = {}, dv: DataViewFacade): IChartConfig {
    const isDualAxis: boolean = get(config, "dualAxis", true);
    const type: VisType = get(config, "type");
    const { secondary_yaxis: secondaryYAxis, ...remainConfig } = config;
    const secondaryMeasuresIds = dv.bucketMeasures(SECONDARY_MEASURES).map(measureId);

    if (!isComboChart(type)) {
        return config;
    }

    if (!isDualAxis) {
        return remainConfig;
    }

    return {
        ...remainConfig,
        secondary_yaxis: {
            ...secondaryYAxis,
            measures: secondaryMeasuresIds,
        },
    };
}
