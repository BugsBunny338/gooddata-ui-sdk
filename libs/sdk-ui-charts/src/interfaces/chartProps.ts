// (C) 2019-2020 GoodData Corporation
import { IAnalyticalBackend, IPreparedExecution } from "@gooddata/sdk-backend-spi";
import * as React from "react";
import {
    IHeaderPredicate,
    IErrorProps,
    ILoadingProps,
    IDrillableItem,
    OnFiredDrillEvent,
    IPushData,
    OnError,
    OnExportReady,
    OnLoadingChanged,
    IVisualizationCallbacks,
    IVisualizationProps,
} from "@gooddata/sdk-ui";
import { IChartConfig } from "./chartConfig";

/**
 * Props applicable for all charts
 *
 * @public
 */
export interface ICommonChartProps extends IVisualizationProps, IChartCallbacks {
    /**
     * Set Locale for chart localization.
     *
     * Note: This locale will be used for everything EXCEPT the data being visualized.
     */
    locale?: string;

    /**
     * Configure chart drillability; e.g. which parts of the charts can be clicked.
     */
    drillableItems?: Array<IDrillableItem | IHeaderPredicate>;

    /**
     * Configure chart's behavior and appearance.
     */
    config?: IChartConfig;

    /**
     * Set height of the chart (in pixels).
     */
    height?: number;

    /**
     * React component to display in the event when there is an error with either obtaining the data from
     * backend or with the data itself.
     */
    ErrorComponent?: React.ComponentType<IErrorProps>;

    /**
     * React component to display while loading data from the backend.
     */
    LoadingComponent?: React.ComponentType<ILoadingProps>;
}

export interface ILegendItem {
    name: string;
    color: string; // hex or RGB, can be used directly in CSS style
    onClick: () => void; // toggle to show/hide serie in chart
}

export interface ILegendData {
    legendItems: ILegendItem[];
}

export type OnLegendReady = (data: ILegendData) => void;

/**
 * Defines callbacks to execute for different events.
 *
 * @public
 */
export interface IChartCallbacks extends IVisualizationCallbacks {
    /**
     * Called when an error occurs while loading data for the chart.
     */
    onError?: OnError;

    /**
     * Called when the chart is ready to be exported.
     */
    onExportReady?: OnExportReady;

    /**
     * Called when loading status of the chart changes - chart starts loading or stops loading
     */
    onLoadingChanged?: OnLoadingChanged;

    /**
     * Called when legend is rendered.
     */
    onLegendReady?: OnLegendReady;

    /**
     * Called when user triggers a drill on a chart.
     */
    onDrill?: OnFiredDrillEvent;

    /**
     * @internal
     */
    afterRender?: () => void;

    /**
     * @internal
     */
    pushData?: (data: IPushData) => void;
}

//
// Prop types extended by the bucket components
//

/**
 * Props for all bucket charts.
 *
 * @public
 */
export interface IBucketChartProps extends ICommonChartProps {
    /**
     * Analytical backend, from which the chart will obtain data to visualize
     *
     * If you do not specify instance of analytical backend using this prop, then you MUST have
     * BackendProvider up in the component tree.
     */
    backend?: IAnalyticalBackend;

    /**
     * Identifier of analytical workspace, from which the chart will obtain data to visualize.
     *
     * If you do not specify workspace identifier, then you MUST have WorkspaceProvider up in the
     * component tree.
     */
    workspace?: string;
}

//
// Props for all core charts
//

/**
 * Props for all Core* charts.
 *
 * NOTE: Core* charts are NOT part of public API.
 *
 * @internal
 */
export interface ICoreChartProps extends ICommonChartProps {
    /**
     * Prepared execution, which when executed, will provide data to visualize in the chart.
     */
    execution: IPreparedExecution;
}
