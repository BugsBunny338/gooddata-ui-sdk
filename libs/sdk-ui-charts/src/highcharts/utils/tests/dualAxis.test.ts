// (C) 2019-2020 GoodData Corporation
import { setMeasuresToSecondaryAxis } from "../dualAxis";
import { IChartConfig } from "../../../interfaces";
import { recordedDataView } from "@gooddata/sdk-backend-mockingbird";
import { ReferenceRecordings } from "@gooddata/reference-workspace";

describe("setMeasuresToSecondaryAxis", () => {
    const ComboChart = recordedDataView(
        ReferenceRecordings.Scenarios.ComboChart.OnePrimaryAndSecondaryMeasureWithViewBySortedByAttr,
    );
    const dualAxisConfig = { secondary_yaxis: { rotation: "90" } };

    const TEST_CASES = [
        ["", true],
        ["", undefined],
    ];

    it.each(TEST_CASES)(
        "should %s add measures to secondary axis when dualAxis=%s",
        (_desc: string, value: boolean) => {
            const config: IChartConfig = {
                type: "combo2",
                dualAxis: value,
                ...dualAxisConfig,
            };
            expect(setMeasuresToSecondaryAxis(config, ComboChart)).toMatchSnapshot();
        },
    );
});
