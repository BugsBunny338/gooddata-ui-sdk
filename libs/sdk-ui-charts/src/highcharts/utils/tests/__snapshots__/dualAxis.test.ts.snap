// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setMeasuresToSecondaryAxis should  add measures to secondary axis when dualAxis=true 1`] = `
Object {
  "dualAxis": true,
  "secondary_yaxis": Object {
    "measures": Array [
      "m_acugFHNJgsBy",
    ],
    "rotation": "90",
  },
  "type": "combo2",
}
`;

exports[`setMeasuresToSecondaryAxis should  add measures to secondary axis when dualAxis=undefined 1`] = `
Object {
  "dualAxis": undefined,
  "secondary_yaxis": Object {
    "measures": Array [
      "m_acugFHNJgsBy",
    ],
    "rotation": "90",
  },
  "type": "combo2",
}
`;
